src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/InMemoryPartitionedAppTokenCacheAccessor.cs

```diff
@@ -0,0 +1,200 @@
+            new ConcurrentDictionary<string, ConcurrentDictionary<string, MsalAccessTokenCacheItem>>(1, 1);
+
+        internal /* internal for test only */  readonly ConcurrentDictionary<string, MsalAppMetadataCacheItem> AppMetadataDictionary =
+           new ConcurrentDictionary<string, MsalAppMetadataCacheItem>(1, 1);
+
+        protected readonly ICoreLogger _logger;
+
+        public InMemoryPartitionedAppTokenCacheAccessor(ICoreLogger logger)
+        {
+            _logger = logger ?? throw new System.ArgumentNullException(nameof(logger));
+        }
+
+        #region Add
+        public void SaveAccessToken(MsalAccessTokenCacheItem item)
+        {
+            var partitionKey = CacheKeyFactory.GetClientCredentialKey(item.ClientId, item.TenantId);
+            string itemKey = item.GetKey().ToString();
+            // if a conflict occurs, pick the latest value
+            AccessTokenCacheDictionary
+                .GetOrAdd(partitionKey, new ConcurrentDictionary<string, MsalAccessTokenCacheItem>())[itemKey] = item;
+        }
+
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no refresh tokens, ID tokens, or accounts in a client credential flow
+        /// </summary>
+        public void SaveRefreshToken(MsalRefreshTokenCacheItem item)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no refresh tokens, ID tokens, or accounts in a client credential flow
+        /// </summary>
+        public void SaveIdToken(MsalIdTokenCacheItem item)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no refresh tokens, ID tokens, or accounts in a client credential flow
+        /// </summary>
+        public void SaveAccount(MsalAccountCacheItem item)
+        {
+            throw new NotSupportedException();
+        }
+
+        public void SaveAppMetadata(MsalAppMetadataCacheItem item)
+        {
+            string key = item.GetKey().ToString();
+            AppMetadataDictionary[key] = item;
+        }
+        #endregion
+
+        #region Get
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no refresh tokens, ID tokens, or accounts in a client credential flow
+        /// </summary>
+        public MsalIdTokenCacheItem GetIdToken(MsalAccessTokenCacheItem accessTokenCacheItem)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no refresh tokens, ID tokens, or accounts in a client credential flow
+        /// </summary>
```

Comment:
there is no GetRefreshToken?

Reply:
That's right, there is no GetAccessToken by key and no GetRefreshToken by key, good observation. This is because the cache item keys do not contain the OBO assertion, so we cannot re-construct the partition key from the item key for them. 

But MSAL never actually needed these methods, it always used GetAllAcessTokens / GetAllRefreshTokens, which can be optimized with partitioning.